dipy/align/imaffine.py

```diff
@@ -0,0 +1,1044 @@
+                                    moving.shape, moving_grid2world)
+
+        if self.dim == 2:
+            self.interp_method = vf.interpolate_scalar_2d
+        else:
+            self.interp_method = vf.interpolate_scalar_3d
+
+        if self.sampling_proportion is None:
+            self.update_pdfs = self.update_pdfs_dense
+            self.update_gradient = self.update_gradient_dense
+            self.samples = None
+            self.ns = 0
+            self.transformed = self.affine_map.transform(self.moving)
+            self.transformed = self.transformed.astype(np.float64)
+        else:
+            self.update_pdfs = self.update_pdfs_sparse
+            self.update_gradient = self.update_gradient_sparse
+            self.transformed = None
+            k = int(np.ceil(1.0 / self.sampling_proportion))
+            shape = np.array(static.shape, dtype=np.int32)
+            self.samples = sample_domain_regular(k, shape, static_grid2world)
+            self.samples = np.array(self.samples)
+            self.ns = self.samples.shape[0]
+            # Add a column of ones (homogeneous coordinates)
+            self.samples = np.hstack((self.samples, np.ones(self.ns)[:, None]))
+            # Sample the static image
+            static_p = self.static_world2grid.dot(self.samples.T).T
+            static_p = static_p[..., :self.dim]
+            self.static_vals, inside = self.interp_method(static, static_p)
+            self.static_vals = np.array(self.static_vals, dtype=np.float64)
+            # Sample the moving image
+            sp_to_moving = self.moving_world2grid.dot(P)
+            moving_p = sp_to_moving.dot(self.samples.T).T
+            moving_p = moving_p[..., :self.dim]
+            self.moving_vals, inside = self.interp_method(moving, moving_p)
+        super(MIMetric, self).setup(self.static, self.moving)
+
+    def _update(self, params, update_gradient=True):
+        r""" Updates marginal and joint distributions and the joint gradient
+
+        The distributions are updated according to the static and transformed
+        images. The transformed image is precisely the moving image after
+        transforming it by the transform defined by the `params` parameters.
+
+        The gradient of the joint PDF is computed only if update_gradient
+        is True.
+
+        Parameters
+        ----------
+        params : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+        update_gradient : Boolean, optional
+            if True, the gradient of the joint PDF will also be computed,
+            otherwise, only the marginal and joint PDFs will be computed.
+            The default is True.
+        """
+        # Get the matrix associated with the params parameter vector
+        M = self.transform.param_to_matrix(params)
```

Comment:
> So, my idea for avoiding distance_and_gradient is something like:...

Yes, sorry about that, many of my decisions have been driven by an attempt of making things faster in cython, which often result in not very elegant solutions =(. There are two ways we save time by calling `distance_and_gradient` instead of `distance` followed by `gradient`:
1. The joint and marginal pdfs are re-computed only once (this can also be achieved with your proposed strategy, by remembering whether we just updated the pdfs or not)
2.  The cython function that makes the actual computation of the metric and gradient, is called only once, here:
https://github.com/omarocegueda/dipy/blob/imaffine_pr3/dipy/align/parzenhist.pyx#L1285
This cannot be achieved by remembering whether we just called `distance` previously or not. We will necessarily call this function twice, once without gradient (called by `distance`, which does not need the gradient) and then again with gradient (called by `gradient`).

I think (1) is much more important than (2) because the cost of calling the cython function twice is dominated by the option "with" gradient. So, your proposed solution should not be significantly slower than what we have now.